File: Go/Main/mergesort.go
package main

func merge(nums []int, start int, mid int, end int) {
	n1 := mid - start + 1
	n2 := end - mid
	l := make([]int, n1)
	r := make([]int, n2)

	// Copy arrays
	copy(l, nums[start:mid+1])
	copy(r, nums[mid+1:end+1])

	i := 0
	j := 0
	k := start

	for i < n1 && j < n2 {
		if l[i] <= r[j] {
			nums[k] = l[i]
			i++
		} else {
			nums[k] = r[j]
			j++
		}
		k++
	}

	for i < n1 {
		nums[k] = l[i]
		i++
		k++
	}

	for j < n2 {
		nums[k] = r[j]
		j++
		k++
	}
}

func merge_sort(nums []int, start int, end int) {
	if start < end {
		mid := start + (end-start)/2
		merge_sort(nums, start, mid)
		merge_sort(nums, mid+1, end)
		merge(nums, start, mid, end)
	}
}

File: Java/MergeSort.java
package Java;

class MergeSort {

  public static void mergeSort(int x[], int start, int end) {
    if(start < end) {
        // Divide
        int mid = start + (end-start) / 2;
        mergeSort(x, start, mid);
        mergeSort(x, mid + 1, end);

        // Merge
        merge(x, start, mid, end);
    }
  }

  public static void merge(int x[], int start, int mid, int end) {
    int i,j,k;
    int n1 = (mid - start) + 1;
    int n2 = end - mid;
    int L[] = new int[n1];
    int R[] = new int[n2];
    
    // Copy arrays
    System.arraycopy(x, start, L, 0, n1); // todo check correctness
    System.arraycopy(x, mid+1, R, 0, n2);

    i = 0; 
    j = 0;
    k = start; 

    while(i < n1 && j < n2) {
        if(L[i] < R[j]) 
            x[k++] = L[i++];
        else 
            x[k++] = R[j++];
    }

    while(i < n1) {
        x[k++] = L[i++];
    }
    while(j < n2) {
        x[k++] = R[j++];
    }
  }
}
